File: scripts/fix-html-utf8.php
<?php
declare(strict_types=1);

$root = dirname(__DIR__);

$googleInter = '/\s*<link rel="preconnect" href="https:\/\/fonts\.googleapis\.com">\s*'
    . '<link rel="preconnect" href="https:\/\/fonts\.gstatic\.com" crossorigin>\s*'
    . '<link rel="stylesheet" href="https:\/\/fonts\.googleapis\.com\/css2\?family=Inter:wght@400;500;600;700&display=swap">\s*/i';

$googleMaterial = '/\s*<link rel="stylesheet" href="https:\/\/fonts\.googleapis\.com\/css2\?family=Material\+Symbols[^>]*>\s*/i';

function autodocs_inter_href(string $rel): string
{
    $depth = substr_count($rel, '/');
    if ($depth === 0) {
        return './estilos/fontes/inter.css';
    }
    return str_repeat('../', $depth) . 'estilos/fontes/inter.css';
}

function autodocs_has_corrupted_accents(string $text): bool
{
    return (bool) preg_match(
        '/Usu\?rios|Documenta\?\?|fun\?\?o|n\?o |p\?gina|Configura\?\?|Cr\?dito|Emiss\?o|Raz\?o|Condi\?\?es|institui\?\?o|CL\?USULA|Presta\?\?o|Servi\?os|opera\?\?o|contrata\?\?o/',
        $text
    );
}

function autodocs_looks_utf8_pt(string $raw): bool
{
    if (!mb_check_encoding($raw, 'UTF-8')) {
        return false;
    }
    return (bool) preg_match('/[áàâãéêíóôõúçÁÀÂÃÉÊÍÓÔÕÚÇ]/u', $raw);
}

$iterator = new RecursiveIteratorIterator(
    new RecursiveDirectoryIterator($root, FilesystemIterator::SKIP_DOTS)
);

$fixed = 0;
foreach ($iterator as $file) {
    /** @var SplFileInfo $file */
    if (!$file->isFile() || strtolower($file->getExtension()) !== 'html') {
        continue;
    }
    $path = $file->getPathname();
    $rel = str_replace('\\', '/', substr($path, strlen($root) + 1));
    if (str_starts_with($rel, 'docs/') || str_contains($rel, '/.git/')) {
        continue;
    }

    $raw = file_get_contents($path);
    if ($raw === false) {
        fwrite(STDERR, "READ FAIL $rel\n");
        continue;
    }

    if (!autodocs_looks_utf8_pt($raw) && !mb_check_encoding($raw, 'UTF-8')) {
        $text = iconv('Windows-1252', 'UTF-8//IGNORE', $raw);
        $enc = 'cp1252';
        if ($text === false) {
            fwrite(STDERR, "ICONV FAIL $rel\n");
            continue;
        }
    } else {
        $text = $raw;
        $enc = 'utf-8';
    }

    // Restore from git if accents were replaced by literal '?'
    if (autodocs_has_corrupted_accents($text)) {
        $git = shell_exec('git -C ' . escapeshellarg($root) . ' show ' . escapeshellarg('HEAD:' . $rel));
        if (is_string($git) && $git !== '') {
            $text = $git;
            $enc = 'git-HEAD';
            echo "RESTORED $rel\n";
        }
    }

    $href = autodocs_inter_href($rel);
    $text2 = preg_replace($googleInter, "\n", $text) ?? $text;
    $text2 = preg_replace($googleMaterial, "\n", $text2) ?? $text2;

    if (!preg_match('/fontes\/inter\.css/i', $text2)) {
        $link = '    <link rel="stylesheet" href="' . $href . '">';
        if (preg_match('/(<link[^>]*rel="icon"[^>]*>)/i', $text2, $m, PREG_OFFSET_CAPTURE)) {
            $pos = $m[0][1] + strlen($m[0][0]);
            $text2 = substr($text2, 0, $pos) . "\n" . $link . substr($text2, $pos);
        } else {
            $text2 = preg_replace('/<\/head>/i', $link . "\n</head>", $text2, 1) ?? $text2;
        }
    } else {
        $text2 = preg_replace(
            '/<link rel="stylesheet" href="[^"]*fontes\/inter\.css">/',
            '<link rel="stylesheet" href="' . $href . '">',
            $text2
        ) ?? $text2;
    }

    $text2 = preg_replace("/\n{3,}/", "\n\n", $text2) ?? $text2;

    if ($text2 !== $raw) {
        file_put_contents($path, $text2);
        $fixed++;
        echo "WROTE ($enc) $rel\n";
    }
}

echo "fixed_files $fixed\n";

$usuarios = file_get_contents($root . '/usuarios/index.html');
if ($usuarios === false || !str_contains($usuarios, 'Usuários')) {
    fwrite(STDERR, "VERIFY FAIL usuarios\n");
    exit(1);
}
$docs = file_get_contents($root . '/documentacoes/index.html');
if ($docs === false || !str_contains($docs, 'Documentações')) {
    fwrite(STDERR, "VERIFY FAIL documentacoes\n");
    exit(1);
}
echo "VERIFY OK\n";

File: scripts/fix-html-utf8.py
# -*- coding: utf-8 -*-
"""Fix HTML encoding (cp1252 -> UTF-8) and ensure local Inter font links."""
from __future__ import annotations

import pathlib
import re
import subprocess

ROOT = pathlib.Path(__file__).resolve().parents[1]

GOOGLE_INTER = re.compile(
    r"\s*<link rel=\"preconnect\" href=\"https://fonts\.googleapis\.com\">\s*"
    r"<link rel=\"preconnect\" href=\"https://fonts\.gstatic\.com\" crossorigin>\s*"
    r"<link rel=\"stylesheet\" href=\"https://fonts\.googleapis\.com/css2\?family=Inter:wght@400;500;600;700&display=swap\">\s*",
    re.I,
)
GOOGLE_MATERIAL = re.compile(
    r"\s*<link rel=\"stylesheet\" href=\"https://fonts\.googleapis\.com/css2\?family=Material\+Symbols[^>]*>\s*",
    re.I,
)
INTER_LINK = re.compile(r"fontes/inter\.css", re.I)
LATIN1_MARKERS = (
    "Usuários".encode("cp1252"),
    "Documentações".encode("cp1252"),
    "Função".encode("cp1252"),
    "não".encode("cp1252"),
    "página".encode("cp1252"),
)


def inter_href(rel_posix: str) -> str:
    depth = rel_posix.count("/")
    if depth == 0:
        return "./estilos/fontes/inter.css"
    return "../" * depth + "estilos/fontes/inter.css"


def load_text(path: pathlib.Path) -> tuple[str, str]:
    raw = path.read_bytes()
    if any(m in raw for m in LATIN1_MARKERS):
        return raw.decode("cp1252"), "cp1252"
    try:
        return raw.decode("utf-8"), "utf-8"
    except UnicodeDecodeError:
        return raw.decode("cp1252"), "cp1252"


def maybe_restore_from_git(path: pathlib.Path, text: str) -> str:
    if not re.search(r"Usu\?rios|Documenta\?\?|fun\?\?o|n\?o |p\?gina", text):
        return text
    rel = path.relative_to(ROOT).as_posix()
    data = subprocess.check_output(["git", "show", f"HEAD:{rel}"], cwd=ROOT)
    print("RESTORED", rel)
    return data.decode("utf-8")


def ensure_inter(text: str, href: str) -> str:
    text = GOOGLE_INTER.sub("\n", text)
    text = GOOGLE_MATERIAL.sub("\n", text)
    if INTER_LINK.search(text):
        text = re.sub(
            r'<link rel="stylesheet" href="[^"]*fontes/inter\.css">',
            f'<link rel="stylesheet" href="{href}">',
            text,
        )
        return text
    link = f'    <link rel="stylesheet" href="{href}">'
    m = re.search(r'(<link[^>]*rel="icon"[^>]*>)', text, re.I)
    if m:
        return text[: m.end()] + "\n" + link + text[m.end() :]
    return text.replace("</head>", link + "\n</head>", 1)


def main() -> None:
    html_files = [
        p
        for p in ROOT.rglob("*.html")
        if "docs" not in p.parts and ".git" not in p.parts
    ]
    fixed = 0
    for path in html_files:
        raw = path.read_bytes()
        text, enc = load_text(path)
        text = maybe_restore_from_git(path, text)
        rel = path.relative_to(ROOT).as_posix()
        href = inter_href(rel)
        text2 = ensure_inter(text, href)
        text2 = re.sub(r"\n{3,}", "\n\n", text2)
        out = text2.encode("utf-8")
        if out != raw:
            path.write_bytes(out)
            fixed += 1
            print(f"WROTE ({enc}) {rel}")
    print("fixed_files", fixed)

    usuarios = (ROOT / "usuarios" / "index.html").read_bytes()
    assert "Usuários".encode("utf-8") in usuarios, "usuarios title not UTF-8"
    docs = (ROOT / "documentacoes" / "index.html").read_bytes()
    assert "Documentações".encode("utf-8") in docs, "documentacoes title not UTF-8"
    print("VERIFY OK")


if __name__ == "__main__":
    main()
